uaclient/log.py

```diff
@@ -58,3 +63,87 @@ def format(self, record: logging.LogRecord) -> str:
 
         local_log_record["extra"] = extra_message_dict
         return json.dumps(list(local_log_record.values()))
+
+
+class LogFile(enum.Enum):
+    MAIN = 0
+    TIMER = 1
+    DAEMON = 2
+
+    def default_file(self) -> str:
+        if self == self.MAIN:
+            return CONFIG_DEFAULTS["log_file"]
+        raise NotImplementedError("Implement me for {}".format(self))
+
+    def extract_log_file(self, ua_cfg: config.UAConfig) -> str:
+        if self == self.MAIN:
+            return ua_cfg.log_file
+        raise NotImplementedError("Implement me for {}".format(self))
+
+
+def setup_logging(console_level, log_level, log_file=None, logger=None):
+    """Setup console logging and debug logging to log_file"""
+    if log_file is None:
+        cfg = config.UAConfig()
+        log_file = cfg.log_file
+    console_formatter = util.LogFormatter()
+    if logger is None:
+        # Then we configure the root logger
+        logger = logging.getLogger()
+    logger.setLevel(log_level)
+    logger.addFilter(RedactionFilter())
+
+    # Clear all handlers, so they are replaced for this logger
+    logger.handlers = []
+
+    # Setup console logging
+    console_handler = logging.StreamHandler(sys.stderr)
+    console_handler.setFormatter(console_formatter)
+    console_handler.setLevel(console_level)
+    console_handler.set_name("ua-console")  # Used to disable console logging
+    logger.addHandler(console_handler)
+
+    # Setup file logging
+    if util.we_are_currently_root():
+        # Setup readable-by-root-only debug file logging if running as root
+        log_file_path = pathlib.Path(log_file)
+
+        if not log_file_path.exists():
+            log_file_path.touch()
+            log_file_path.chmod(0o644)
+
+        file_handler = logging.FileHandler(log_file)
+        file_handler.setFormatter(JsonArrayFormatter())
+        file_handler.setLevel(log_level)
+        file_handler.set_name("ua-file")
+        logger.addHandler(file_handler)
+
+
+class EarlyLoggingSetup:
```

Comment:
The reasons to implement this as a context manager are:

* Perform the initial log config in the `__enter__`.
* During the lifetime of the CM, client code could potentially need the uacfg. If so, they can hand over it by setting `self.ua_cfg` and avoid multiple reads of the config.
* On __exit__ time, the CM will set up the final logging config.
* If, in the future, more state needs to be tracked between those two events, a CM is easier to maintain than two calls to free functions.

An example of usage: https://github.com/canonical/ubuntu-pro-client/pull/2486/files#diff-d558bdbcb7ee1bef85fd138367f1e64708ea8271c09cf8e4c19941c24c8883d6R1923

Reply:
I like the idea, but I'm worried it's a premature design. IOW, I agree that "If more state needs to be tracked..." then this would be nice, but right now we don't have any extra state. So the simplicity of directly calling setup_logging twice with different parameters is appealing.
Maybe I'm missing other advantages though - what do you think?

Reply:
Yeah, I do agree, probably overengineering here. I am going to close this PR and I will re-open it with a simplified version, thanks!